### PR TITLE
e2e: fix: get kubelet port from node status instead of hard coding

### PR DESCRIPTION
**Pull request by zhuangqh**
Signed-off-by: zhuangqh <zhuangqhc@gmail.com>

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change

/kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

get kubelet port from node status instead of hard coding. Not all the kubelet port is 10250.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Review by alejandrox1 — COMMENTED** [review]
Thank you for catching this @zhuangqh !
/lgtm
/ok-to-test
/priority backlog
/kind cleanup

/cc @kubernetes/sig-node-pr-reviews 
ptal

**Comment by liggitt** [comment]
Anywhere you are intending to proxy to the kubelet and are currently hard-coding the kubelet port, I would suggest dropping the port altogether. The default when proxying to a node is to proxy to the kubelet port, which the kube-apiserver will look up from the specific kubelet's reported status.

See https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/node/strategy.go#L240-L255 and https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/client/kubelet_client.go#L209-L226

File: test/e2e/framework/kubelet/kubelet_pods.go [hunk 1]
```diff
@@ -17,12 +17,23 @@ limitations under the License.
...
 	"k8s.io/kubernetes/test/e2e/framework"
 )
 
+// GetKubeletPort find the kubelet port from the given nodeName.
+func GetKubeletPort(c clientset.Interface, nodeName string) (int, error) {
+	node, err := c.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
+	if err != nil {
+		return -1, fmt.Errorf("Failed to get node %s: %v", nodeName, err)
+	}
+	return int(node.Status.DaemonEndpoints.KubeletEndpoint.Port), nil
```
Comment (timothysc) [inline test/e2e/framework/kubelet/kubelet_pods.go]: Why is the casting here necessary, what is the returned type?

**Review by timothysc — APPROVED** [review]
I don't mind the change, but is are you changing the ports in your environment? 

/approve 
/hold

File: test/e2e/framework/kubelet/kubelet_pods.go (hunk 1 above)
Comment (zhuangqh) [reply test/e2e/framework/kubelet/kubelet_pods.go]: @timothysc original type is `int32`

**Comment by zhuangqh** [comment]
> I don't mind the change, but is are you changing the ports in your environment?

@timothysc yep. I am working on a [multi-tenancy](https://github.com/kubernetes-sigs/multi-tenancy/tree/master/incubator/virtualcluster) project. There is a fake agent instead of kubelet. The agent's port is different from the real kubelet

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/86071#issuecomment-566179825 is still relevant... letting the apiserver do the node port lookup would be more reliable and simpler

**Comment by zhuangqh** [comment]
> [#86071 (comment)](https://github.com/kubernetes/kubernetes/pull/86071#issuecomment-566179825) is still relevant... letting the apiserver do the node port lookup would be more reliable and simpler

I totally agree with you. But I still not figure out how to do the node proxy as you said😂
could you give me more input

**Comment by liggitt** [comment]
> But I still not figure out how to do the node proxy as you said😂
> could you give me more input. I am not familiar with k8s code😅

the simplest approach I can think of is to accept a port of `0` as a special case, and in ProxyRequest, if the port is zero, construct the proxy request to `Name(node)`

**Comment by liggitt** [comment]
then you no longer need `GetKubeletPort`

**Comment by zhuangqh** [comment]
> > But I still not figure out how to do the node proxy as you said😂
> > could you give me more input. I am not familiar with k8s code😅
> 
> the simplest approach I can think of is to accept a port of `0` as a special case, and in ProxyRequest, if the port is zero, construct the proxy request to `Name(node)`

good idea. I am trying

**Pull request closed without merging**